Cache results when checking for valid attributes

areveny opened the pull request:
## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Early development to give reference for discussion.

Closes #5695

Pierre-Sassoulas commented on pylint/checkers/typecheck.py:
```diff
@@ -1013,6 +1016,24 @@ def visit_attribute(self, node: nodes.Attribute) -> None:
+            outer = expr.attrname + outer
+        name_with_expr = f"{outer}.{node.attrname}"
+        if (
+            outer != ""
```
```suggestion
            outer

Pierre-Sassoulas commented on pylint/checkers/typecheck.py:
```diff
@@ -1013,6 +1016,24 @@ def visit_attribute(self, node: nodes.Attribute) -> None:
+        name_with_expr = f"{outer}.{node.attrname}"
+        if (
+            outer != ""
+            and name_with_expr in self.node_exists
+            and self.node_exists[name_with_expr]
```
```suggestion
            and self.node_exists.get(name_with_expr) 
```

Pierre-Sassoulas commented on pylint/checkers/typecheck.py:
```diff
@@ -1013,6 +1016,24 @@ def visit_attribute(self, node: nodes.Attribute) -> None:
 
         function/method, super call and metaclasses are ignored
         """
+
```
I think it would be a good thing to isolate the cache updating in a specifically named function si it's explicit that this part is about caching.

Pierre-Sassoulas reviewed (commented):
I liked the analysis, 2 to 7 percent better performance is huge too.

areveny commented:
Added the scope in the cache key so existing tests pass. The performance improvement is smaller but still good, on Django we hit the cache 60% of the time and save 5% runtime.

Currently working on some unittests to try and break this and ensure caches hits/misses are correct.

areveny commented:
Incorporated existing feedback, added unit tests and a ChangeLog message in a provisional place since this is targeting 2.14.

I think there are obscure cases where the cache breaks, for example conditional imports/class definitions in `if` statements, but I am not sure if we would/can pick those up now.

Pierre-Sassoulas commented:
> I think there are obscure cases where the cache breaks

By "break" do you mean "a cache miss", or "a crash" ? 😄

By the way, I should have asked during my first review (sorry), but any reason why we are not using ``functool.lru_cache`` ?

areveny commented:
Break meaning a false negative.

Below using the cache hides an error. `counter.b` is used twice in the same function. The cache assumes that all attributes within a scope with the same name have the same result. The cache returns that `counter.b` is valid in the second access, the current code infers on each access to ensure it is valid to catch this situation.

```
class RootCounter:
    elements = [1, 2, 3]
    b = [1, 2, 3]

class RootCounter2:
    elements = [1, 2, 3]

class CounterContainer1:

    counter = RootCounter()

    def fn(self):
        print(counter.elements)
        print(counter.b)
        counter = RootCounter2()
        print(counter.b) # no-member
```



I'm not using a cache from `functools` because the function accepts a different Attribute node object each time so it wouldn't hit the cache. We could write a custom decorator to pull out the cache key from the node but that's essentially what's happening now.

Also I don't think cache eviction is necessary, the memory impact of storing this cache map should be small compared to the AST.

DanielNoord commented:
@areveny I kinda forgot what the status of this PR is. Is it ready for review? Or is the issue mentioned above still there?

Pierre-Sassoulas commented:
> making assumptions about attributes staying either valid or invalid in a given scope 

Do you have an example where it's not the case or detail the reasoning here in case someone else want to dig into it ? I don't have an example of the validity of an attribute changing inside a scope right now 😄

areveny commented:
Yeah, consider this:

counter.b is used twice in the same function. The cache assumes that all attributes within a scope with the same name have the same result. The cache returns that counter.b is valid in the second access, the current code infers on each access to ensure it is valid to catch this situation.

```
class RootCounter:
    elements = [1, 2, 3]
    b = [1, 2, 3]

class RootCounter2:
    elements = [1, 2, 3]

class CounterContainer1:

    counter = RootCounter()

    def fn(self):
        print(counter.elements)
        print(counter.b)
        counter = RootCounter2()
        print(counter.b) # no-member
```
